tests/e2e/default/sms/rapidpro.wdio-spec.js

```diff
@@ -0,0 +1,593 @@
+    });
+  }); // End describe 'Webapp Terminating messages'
+
+  describe('Webapp originating messages and state updates', () => {
+    const OUTGOING_KEY = 'ermahgerd';
+    const setOutgoingKey = () => {
+      return utils.saveCredentials('rapidpro:outgoing', OUTGOING_KEY);
+    };
+
+    const verifyPhoneAndMessage = (obj, index) => {
+      expect(obj.urns).to.eql([`tel:phone${index}`]);
+      expect(obj.text).to.equal(`message${index}`);
+    };
+
+    const verifyGatewayRefAndState = (obj, gateway, state) => {
+      expect(obj.gateway_ref).to.equal(gateway);
+      expect(obj.state).to.equal(state);
+    };
+
+    const pregnancyReportWithTasks = pregnancyReportFactory.pregnancy().build({
+      tasks: [
+        {
+          messages: [{ to: 'phone1', message: 'message1', uuid: 'uuid1' }],
+          state: 'pending',
+        },
+        {
+          messages: [{ to: 'phone2', message: 'message2', uuid: 'uuid2' }],
+          state: 'sent',
+        },
+      ],
+      scheduled_tasks: [
+        {
+          messages: [{ to: 'phone3', message: 'message3', uuid: 'uuid3' }],
+          state: 'pending',
+          group: 2,
+          type: 'ANC Reminders LMP',
+        },
+        {
+          messages: [{ to: 'phone4', message: 'message4', uuid: 'uuid4' }],
+          state: 'pending',
+          group: 2,
+          type: 'ANC Reminders LMP',
+        },
+        {
+          messages: [{ to: 'phone5', message: 'message5', uuid: 'uuid5' }],
+          state: 'pending',
+          group: 3,
+          type: 'ANC Reminders LMP',
+        },
+      ]
+    });
+
+    beforeEach(async () => {
+      const settings = {
+        sms: {
+          outgoing_service: 'rapidpro',
+          rapidpro: {url: utils.hostURL(server.address().port)},
+        }
+      };
+      await utils.updateSettings(settings, true);
+    });
+
+    afterEach(() => utils.revertDb([], true));
+
+
+    it('should not call RapidPro endpoint when credentials are not set', async () => {
+      await utils.saveDoc(pregnancyReportWithTasks);
+      await browser.pause(1200); // interval to check the queue should run every second
+      expect(broadcastsEndpointRequests.length).to.equal(0);
+    });
+
+    it('should call endpoint with set credentials', async () => {
+      await setOutgoingKey();
+
+      await utils.saveDoc(pregnancyReportWithTasks);
+      await browser.waitUntil(() => broadcastsEndpointRequests.length === 4, 1200);
+
+      const bodies = broadcastsEndpointRequests
+        .map(item => item[0])
+        .sort((a, b) => a.text.localeCompare(b.text));
+
+      verifyPhoneAndMessage(bodies[0], 1);
+      verifyPhoneAndMessage(bodies[1], 3);
+      verifyPhoneAndMessage(bodies[2], 4);
+      verifyPhoneAndMessage(bodies[3], 5);
+
+      const headers = broadcastsEndpointRequests.map(item => item[1]);
+      headers.forEach(header => expect(header.authorization).to.equal(`Token ${OUTGOING_KEY}`));
```

Comment:
I like how you combined all these duplicate assertions into one here with the `forEach`, but you should probably also assert the `headers.length` too (because this line will not fail if `headers` is empty...)

Reply:
That is a really good point. thank you! 👍